Content.Shared/Weapons/Reflect/ReflectSystem.cs

```diff
@@ -66,43 +66,86 @@ private void OnReflectUserHitscan(EntityUid uid, ReflectUserComponent component,
         if (args.Reflected)
             return;
 
-        foreach (var ent in _inventorySystem.GetHandOrInventoryEntities(uid, SlotFlags.WITHOUT_POCKET))
+        EntityUid? reflectorUid = null;
+        ReflectComponent? bestReflector = null;
+
+        foreach (var entityUid in _inventorySystem.GetHandOrInventoryEntities(uid, SlotFlags.WITHOUT_POCKET))
         {
-            if (!TryReflectHitscan(uid, ent, args.Shooter, args.SourceItem, args.Direction, out var dir))
+            if (!TryComp<ReflectComponent>(entityUid, out var comp))
+                continue;
+
+            if (!comp.Enabled)
+                continue;
+
+            if (bestReflector != null && bestReflector.ReflectProb >= comp.ReflectProb)
                 continue;
 
-            args.Direction = dir.Value;
-            args.Reflected = true;
-            break;
+            bestReflector = comp;
+            reflectorUid = entityUid;
         }
+
+        if (reflectorUid == null || bestReflector == null)
+            return;
+
+        if (!TryReflectHitscan(uid, reflectorUid.Value, args.Shooter, args.SourceItem, args.Direction, out var dir))
+            return;
+
+        args.Direction = dir.Value;
+        args.Reflected = true;
     }
 
     private void OnReflectUserCollide(EntityUid uid, ReflectUserComponent component, ref ProjectileReflectAttemptEvent args)
     {
-        foreach (var ent in _inventorySystem.GetHandOrInventoryEntities(uid, SlotFlags.WITHOUT_POCKET))
+        if (args.Cancelled)
+            return;
+
+        EntityUid? reflectorUid = null;
+        ReflectComponent? bestReflector = null;
+
+        if (!TryComp<ReflectiveComponent>(args.ProjUid, out var reflective))
+            return;
+
+        // Get the best reflector in the hit entity's inventory for the projectile's hit
+        foreach (var entityUid in _inventorySystem.GetHandOrInventoryEntities(uid, SlotFlags.WITHOUT_POCKET))
         {
-            if (!TryReflectProjectile(uid, ent, args.ProjUid))
+            if (!TryComp<ReflectComponent>(entityUid, out var comp))
                 continue;
 
-            args.Cancelled = true;
-            break;
+            if (!comp.Enabled)
+                continue;
+
+            if ((comp.Reflects & reflective.Reflective) == 0x0)
+                continue;
+
+            if (bestReflector != null && bestReflector.ReflectProb >= comp.ReflectProb)
+                continue;
+
+            bestReflector = comp;
+            reflectorUid = entityUid;
         }
```

Comment:
these could just be a helper method

Reply:
Yeah there's code tidyup possible in this system, it's quite grotty

Reply:
i mean no reason to expand the problem in this case since i think making a helper method for the loop and checks *should* be easy unless im missing something